feat: add Git (#2627)
OI-wiki/OI-wiki: docs/tools/git.md

???+note

>     本页面将着重介绍 Git 这一版本控制系统，与 GitHub 相关的内容，请参考 [GitHub 帮助](https://docs.github.com/cn) 和 [如何参与 - OI Wiki](https://oi-wiki.org/intro/htc/) 。

Reviewer: ```suggestion
    本页面将着重介绍 Git 这一版本控制系统，与 GitHub 相关的内容，请参考 [GitHub 帮助](https://docs.github.com/cn) 和 [如何参与 - OI Wiki](../intro/htc.md) 。
```
建议煲汤（

Author: 我谢罪 /wq

Git 是目前使用最广泛的版本控制系统。 **OI Wiki** 也使用了 Git 作为版本控制系统。

## 安装

对于 Windows 系统，可以在 [Git 官网](https://git-scm.com/downloads) 获取到 Git 的安装包。

对于 Linux 系统，可以通过包管理器进行安装。不同 Linux 发行版的安装方式可能不同， [Git 官网](https://git-scm.com/download/linux) 给出了不同系统下的安装方式。

> 对于 Mac 系统，只需在终端执行 `brew install git` 命令即可安装。

Reviewer: ```suggestion
对于 macOS 系统，只需在终端执行 `brew install git` 命令即可安装。（需安装 [Homebrew](https://brew.sh/)）
```

Reviewer: 修改被覆盖了（

Reviewer 2: 重新开这个的 review 吧

Reviewer: 我选择直接改（）

## 配置

Git 的配置文件，主要存储在三个位置：

1.  `/etc/gitconfig` ：该配置文件对系统上的所有用户，以及该系统上的所有仓库均有效。
2.  `~/.gitconfig` ：该配置文件对系统的当前用户，以及该系统上的所有仓库有效。
3. 当前仓库目录下的 `.git/config` 文件：只适用于该仓库。

局部设置会自动覆盖全局设置。因此如果需要在某个仓库应用特定的设置的话，只需更改该仓库下的特定设置即可，不会对全局设置造成影响。

修改配置文件需要用到 `git config` 命令。

### 设置用户信息

安装 Git 后，第一件事情就是设置你的用户名和邮箱。这些信息在每次提交时都会用到。

```bash

> $ git config --global user.name "OI Wiki"

Reviewer: 这里要不要标一下改成自己的用户名和邮箱（雾）

Reviewer 2: 我认为还是应该改的，毕竟真的会有照着改成 OI Wiki 的人

$ git config --global user.email hi@oi-wiki.org
```

这里的 `--global` 表示修改的是 `~/.gitconfig` 文件，即该设置对当前用户下的所有仓库均有效。

如果想要修改某个仓库的特定设置，只需在该仓库下执行不带 `--global` 的命令即可。

### 配置编辑器

```bash
$ git config --global core.editor emacs
```

执行如上命令可以将编辑器更改为 [Emacs](./editor/emacs.md) 。

> 在 Windows 下，配置编辑器的过程可能会稍有麻烦，具体配置方式可以在网络上查找相关资料。

Reviewer 3: 为啥会麻烦啊 /yun 是说可执行文件不在 path 里吗。

### 显示配置

可以通过 `git config --list` 列出当前已经设置的所有配置参数。

## 仓库操作基础

### 新建 Git 仓库

> 新建一个 Git 仓库非常简单，只需在你想要建立仓库的文件夹输入如下命令：

Reviewer 2: ```suggestion
新建一个 Git 仓库非常简单，只需在当前目录输入如下命令：
```

Reviewer 3: 不是很懂为啥要改成“当前目录”。我好像在 Reddit 见过在 Windows 的 home 建仓库的（（（

Reviewer 2: 为了避开第一人称啦（

Reviewer 2: 我还是再改改吧

Reviewer 3: 首先这是第二人称。

然后我觉得这样写没啥问题。

你觉得不行的话把”你“去掉也行，”想要建立仓库的文件夹“这个意思要保留。

Reviewer 2: 嗯我直接把“你”去掉了。
之前改掉的原因是 git init 就是直接在当前目录下 init 的，要指定目录的话得在后面加上路径。

Reviewer 3: > 之前改掉的原因是 git init 就是直接在当前目录下 init 的，要指定目录的话得在后面加上路径。

那是命令的作用是“在当前目录新建仓库”，你需要做的事情是“在想要建立仓库的文件夹输入命令”。

```bash
$ git init
```

Git 将在当前文件夹新建一个 `.git` 文件夹，一个仓库就这样建好了。

> 如果别人已经建立了一个仓库，你想要把这个仓库拷贝到自己的电脑上（比如将 **OI Wiki** 的代码拷贝到本地上进行编辑），采用 `git clone` 命令即可。

Reviewer 2: ```suggestion
如果想把别人的仓库拷贝到自己的电脑上（比如将 **OI Wiki** 的代码拷贝到本地上进行编辑），采用 `git clone` 命令即可。
```

Reviewer 3: “把一个仓库拷贝到本地”会不会更好，因为不一定是别人的。

Reviewer 2: 有道理，我改改

```bash
$ git clone https://github.com/OI-wiki/OI-wiki.git
```

这样在当前文件夹下，就会新建一个名为 `OI-wiki` 的仓库，里面存放着原来仓库的所有信息。

### 跟踪文件

在我们对仓库做出了一些修改后，我们需要将这些修改纳入版本管理当中去。

使用 `git status` 命令可以查看当前仓库文件的状态。

假设我们在一个空仓库中新增了一个 `README.md` 文件，执行 `git status` 命令的效果如下：

```bash
$ mkdir test
$ cd test
$ git init # 在 test 文件夹下建立一个新仓库
$ vi README.md # 写点东西
$ git status
On branch master

No commits yet

Untracked files:
  (use "git add <file>..." to include in what will be committed)

        README.md

nothing added to commit but untracked files present (use "git add" to track)
```

这里的 Untracked files 指的是 Git 之前没有纳入版本跟踪的文件。如果文件没有纳入版本跟踪，我们对该文件的修改不会被 Git 记录。

执行 `git add <文件>` 命令可以将指定的文件纳入到版本跟踪中。

```bash
$ git add README.md # 将这个文件纳入到版本跟踪中
$ git status
On branch master

No commits yet

Changes to be committed:
  (use "git rm --cached <file>..." to unstage)

        new file:   README.md
```

这时 `README.md` 已经纳入了版本跟踪，放入了暂存区。我们接下来只需执行 `git commit` 命令就可以提交这次更改了。

但在我们干这个之前，先让我们再对 `README.md` 做点小修改。

```bash

> $ vi README.md # 再随便修改点东西

Reviewer: ```suggestion
$ vim README.md # 再随便修改点东西
```
其他同

接下来两行则详细说明了本次更新涉及的文件修改。

另外，commit 过程中可以利用几个参数来简化提交过程：

-  `-a` ：自动将非暂存区内的提交放入暂存区，然后提交暂存区中的所有更改。需要注意的是未被跟踪的文件（新创建的文件）不会被自动加入暂存区，需要用 `git add` 命令手动添加。
-  `-m` ：该参数后跟提交信息，表示以该提交信息提交本次更改。

### 查看提交记录

使用 `git log` 命令可以查看我们的提交历史记录[^note1]。

可以看到，提交历史里记录了每次提交时的 SHA-1 校验和，提交的作出者，提交时间和 commit 信息。

```bash
$ git log
commit cf78ab3c198337acf508d90924328d476c2e5783 (HEAD -> master)
Author: OI Wiki <hi@oi-wiki.org>
Date:   Fri May 10 22:39:27 2019 +0800

    update README.md

commit b13c84ee799c0fc4c3420814d95c03ea1823ebe3
Author: OI Wiki <hi@oi-wiki.org>
Date:   Fri May 10 22:25:56 2019 +0800

    test
```

### 分支管理

为什么版本管理中需要分支管理呢？答案主要有两点：

1. 直接修改主分支不仅会使历史记录混乱，也可能会造成一些危险的后果。
2. 通过分支，我们可以专注于当前的工作。如果我们需要完成两个不同的工作，只需开两个分支即可，两个分支间的工作互不干扰。

利用 `git branch` 命令可以创建分支， `git checkout` 命令可以切换分支。

```bash
$ git branch dev # 创建一个叫做 dev 的新分支
$ git checkout dev # 切换当前分支到 dev
Switched to branch 'dev'
$ git branch # 查看所有分支信息
  master
* dev
```

???+note "更快速地切换分支"
    上面的操作需要执行两条命令，事实上只需要一条命令就够了。
    
    执行 `git checkout -b dev` 就可以在创建 dev 分支的同时，将当前分支切换到 dev。

dev 前面的星号代表我们当前的分支为 dev，我们的修改都将记录在这个分支上。

我们可以在这个分支上做点小修改，比如创建一个新文件 `test.cpp` 。

```bash
$ vi test.cpp
$ git add test.cpp

> $ git commit -m "QAQ"

Reviewer 2: ```suggestion
$ git commit -m "添加一个 cpp 文件"
```

Reviewer 3: > 一定要让新人从零开始养成良好的 commit 习惯new_moon_with_face

草，“一个 cpp 文件”也算良好吗（

```suggestion
$ git commit -m "添加 test.cpp"
```

其实最好是说一下具体做了什么，但这里没有 test.cpp 的内容，所以只能文件名了。

???+note
    本页面将着重介绍 Git 这一版本控制系统，与 GitHub 相关的内容，请参考 [GitHub 帮助](https://docs.github.com/cn) 和 [如何参与 - OI Wiki](https://oi-wiki.org/intro/htc/)。

> Git 是目前使用最广泛的版本控制系统。**OI Wiki** 也使用了 Git 作为版本控制系统。

Reviewer 3: > Git 是目前使用最广泛的版本控制系统。

来源请求 /fad

Reviewer 4: ???这个也要来源请求？

Reviewer 4: > The Eclipse Foundation reported in its annual community survey that as of May 2014, Git is now the most widely used source-code management tool, with 42.9% of professional software developers reporting that they use Git as their primary source-control system[82] compared with 36.3% in 2013, 32% in 2012; or for Git responses excluding use of GitHub: 33.3% in 2014, 30.3% in 2013, 27.6% in 2012 and 12.8% in 2011.[83] Open-source directory Black Duck Open Hub reports a similar uptake among open-source projects.[84]

https://en.wikipedia.org/wiki/Git

## 安装

对于 Windows 系统，可以在 [Git 官网](https://git-scm.com/downloads) 获取到 Git 的安装包。

对于 Linux 系统，可以通过包管理器进行安装。不同 Linux 发行版的安装方式可能不同，[Git 官网](https://git-scm.com/download/linux) 给出了不同系统下的安装方式。

对于 Mac 系统，只需在终端执行 `brew install git` 命令即可安装。

## 配置

Git 的配置文件，主要存储在三个位置：

1. `/etc/gitconfig`：该配置文件对系统上的所有用户，以及该系统上的所有仓库均有效。
2. `~/.gitconfig`：该配置文件对系统的当前用户，以及该系统上的所有仓库有效。

> 3. 当前仓库目录下的 `.git/config` 文件：只适用于该仓库。

Reviewer 3: 1. 这个是针对 Linux 的，感觉不太好。
2. 一般不需要关心放哪吧，尤其是初学，会 `git config --global` 就够了。

Reviewer 3: unresolved

局部设置会自动覆盖全局设置。因此如果需要在某个仓库应用特定的设置的话，只需更改该仓库下的特定设置即可，不会对全局设置造成影响。

修改配置文件需要用到 `git config` 命令。

### 设置用户信息

安装 Git 后，第一件事情就是设置你的用户名和邮箱。这些信息在每次提交时都会用到。

```bash
$ git config --global user.name "OI Wiki"
$ git config --global user.email hi@oi-wiki.org
```

这里的 `--global` 表示修改的是 `~/.gitconfig` 文件，即该设置对当前用户下的所有仓库均有效。

如果想要修改某个仓库的特定设置，只需在该仓库下执行不带 `--global` 的命令即可。

### 配置编辑器

```bash
$ git config --global core.editor emacs
```

执行如上命令可以将编辑器更改为 [Emacs](./editor/emacs.md)。

在 Windows 下，配置编辑器的过程可能会稍有麻烦，具体配置方式可以在网络上查找相关资料。

### 显示配置

> 可以通过 `git config --list` 列出当前已经设置的所有配置参数。

Reviewer 3: ```suggestion
可以通过 `git config -l` 列出当前已经设置的所有配置参数。使用 `git config --global -l` 可以列出所有全局配置。
```

## 仓库操作基础

### 新建 Git 仓库

新建一个 Git 仓库非常简单，只需在你想要建立仓库的文件夹输入如下命令：

```bash
$ git init
```

Git 将在当前文件夹新建一个 `.git` 文件夹，一个仓库就这样建好了。

如果别人已经建立了一个仓库，你想要把这个仓库拷贝到自己的电脑上（比如将 **OI Wiki** 的代码拷贝到本地上进行编辑），采用 `git clone` 命令即可。

```bash

> $ git clone https://github.com/OI-wiki/OI-wiki.git

Reviewer 3: ```suggestion
$ git clone https://github.com/OI-wiki/OI-wiki
```

```

> 这样在当前文件夹下，就会新建一个名为 `OI-wiki` 的仓库，里面存放着原来仓库的所有信息。

Reviewer 3: ```suggestion
这样在当前文件夹下，就会新建一个名为 `OI-wiki` 的仓库，里面存放着被克隆的仓库的内容。
```

诸如 PR/issues，被 force-push 掉的 commits 之类的应该不会有。

### 跟踪文件

在我们对仓库做出了一些修改后，我们需要将这些修改纳入版本管理当中去。

使用 `git status` 命令可以查看当前仓库文件的状态。

假设我们在一个空仓库中新增了一个 `README.md` 文件，执行 `git status` 命令的效果如下：

```bash

> $ mkdir test
> $ cd test
> $ git init # 在 test 文件夹下建立一个新仓库
> $ vi README.md # 写点东西

Reviewer 3: ```suggestion
```

上文已经“假设我们在一个空仓库中新增了一个 `README.md` 文件”了，这里有点冗余。

No commits yet

Changes to be committed:
  (use "git rm --cached <file>..." to unstage)

        new file:   README.md

```

这时 `README.md` 已经纳入了版本跟踪，放入了暂存区。我们接下来只需执行 `git commit` 命令就可以提交这次更改了。

但在我们干这个之前，先让我们再对 `README.md` 做点小修改。

```bash
$ vi README.md # 再随便修改点东西
$ git status
On branch master

No commits yet

Changes to be committed:
  (use "git rm --cached <file>..." to unstage)

        new file:   README.md

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

        modified:   README.md
```

你会发现 `README.md` 同时处于暂存区和非暂存区。如果这时候执行 `git commit` 命令，只有处于暂存区的更改会被提交，而非暂存区的修改，则不会被提交。

git 的提示告诉我们，执行 `git add README.md` 就可以将非暂存区的更改放入暂存区了。

现在我们将非暂存区的文件加入暂存区，将所有更改一并提交。

```bash
$ git add README.md
$ git commit # 接下来会弹出编辑器页面，你需要写下 commit 信息
[master (root-commit) b13c84e] test
 1 file changed, 2 insertions(+)
 create mode 100644 README.md
```

我们现在重点观察一下 commit 的信息。

`master` 表示当前位于 `master` 分支（关于分支的问题，下文将会详细介绍），`b13c84e` 表示本次提交的 SHA-1 校验和的前几位，后面则是本次提交的信息。

需要特别关注的是这里的 SHA-1 校验码，每个校验码都与某个时刻仓库的一个快照相对应。利用这一特性我们可以访问历史某个时刻的仓库快照，并在该快照上进行更改。

接下来两行则详细说明了本次更新涉及的文件修改。

> 另外，commit 过程中可以利用几个参数来简化提交过程：

Reviewer 3: 在合适的位置加个 `git add -A` 吧。

Author: 这个命令是啥 🤔，没在文档里找到 /fad

Reviewer 3: > 这个命令是啥 🤔，没在文档里找到 /fad

添加所有未被忽略的文件，包括 untracked。你用的什么版本，为啥会找不到啊..

Reviewer 4: ```
❯ git add -h | grep A
    -A, --all             add changes from all tracked and untracked files
```

Reviewer 3: unresolved

Author: 啊之前看成 `git commit -A` 了，我谢罪（捂脸

> - `-a`：自动将非暂存区内的提交放入暂存区，然后提交暂存区中的所有更改。需要注意的是未被跟踪的文件（新创建的文件）不会被自动加入暂存区，需要用 `git add` 命令手动添加。

Reviewer 3: ```suggestion
- `-a`：在提交前将所有已跟踪的文件的更改放入暂存区。需要注意的是未被跟踪的文件（新创建的文件）不会被自动加入暂存区，需要用 `git add` 命令手动添加。
```

- `-m`：该参数后跟提交信息，表示以该提交信息提交本次更改。

### 查看提交记录

使用 `git log` 命令可以查看我们的提交历史记录[^note1]。

> 可以看到，提交历史里记录了每次提交时的 SHA-1 校验和，提交的作出者，提交时间和 commit 信息。

Reviewer 3: ```suggestion
可以看到，提交历史里记录了每次提交时的 SHA-1 校验和，提交的作者，提交时间和 commit 信息。
```

这里确实是 author 而不是 committer，例如 cherry-pick 后 author 不变。

```bash
$ git log
commit cf78ab3c198337acf508d90924328d476c2e5783 (HEAD -> master)
Author: OI Wiki <hi@oi-wiki.org>
Date:   Fri May 10 22:39:27 2019 +0800

    update README.md

commit b13c84ee799c0fc4c3420814d95c03ea1823ebe3
Author: OI Wiki <hi@oi-wiki.org>
Date:   Fri May 10 22:25:56 2019 +0800

    test
```

### 分支管理

为什么版本管理中需要分支管理呢？答案主要有两点：

1. 直接修改主分支不仅会使历史记录混乱，也可能会造成一些危险的后果。
2. 通过分支，我们可以专注于当前的工作。如果我们需要完成两个不同的工作，只需开两个分支即可，两个分支间的工作互不干扰。

> 利用 `git branch` 命令可以创建分支，`git checkout` 命令可以切换分支。

Reviewer 3: ```suggestion
利用 `git branch` 命令可以创建分支，`git switch` 命令可以切换分支，`git switch -c` 命令可以创建分支并切换到这个新分支。
```

Reviewer 3: checkout 已经是时代的眼泪了（两个功能分别被 `git switch` 和 `git restore` 取代了）（虽然我自己还在用 checkout）

Author: 学多.webp

```bash

> $ git branch dev # 创建一个叫做 dev 的新分支
> $ git checkout dev # 切换当前分支到 dev

Reviewer 3: ```suggestion
$ git switch -c dev # 创建一个叫做 dev 的新分支并切换当前分支到 dev
```

Switched to branch 'dev'

> $ git branch # 查看所有分支信息

Reviewer 3: ```suggestion
$ git branch # 查看分支列表
```

远程分支等不会在这显示，而且这里只有名字没有其它信息。

  master
* dev
```

> ???+note "更快速地切换分支"
>     上面的操作需要执行两条命令，事实上只需要一条命令就够了。
>     
>     执行 `git checkout -b dev` 就可以在创建 dev 分支的同时，将当前分支切换到 dev。
> 

Reviewer 3: ```suggestion
```

大多数情况下创建分支都会立刻切过去，不用放 note 里吧。

Reviewer 2: 不，并不会，直接用 `git branch` 命令的话，git 是不会切到新分支去的

Reviewer 3: > 不，并不会，直接用 `git branch` 命令的话，git 是不会切到新分支去的

我是说，大多数情况你会切过去，不是大多数命令会切过去。

dev 前面的星号代表我们当前的分支为 dev，我们的修改都将记录在这个分支上。

我们可以在这个分支上做点小修改，比如创建一个新文件 `test.cpp`。

```bash
$ vi test.cpp
$ git add test.cpp

> $ git commit -m "QAQ"

Reviewer 3: 不要 QAQ，不要 QAQ，不要 QAQ（

[dev 4fe4923] QAQ
 1 file changed, 8 insertions(+)
 create mode 100644 test.cpp
```

> 在新分支上修改似乎和在 master 分支上修改没什么太大区别，事实上也确实如此，每个分支都是平等的。

Reviewer 3: ```suggestion
```

理论上平等，实践中不平等。

## 参考资料与注释

> [^note1]: `git log` 的更多用法可以参考 [Pro Git Book v2](https://git-scm.com/book/zh/v2/Git-%E5%9F%BA%E7%A1%80-%E6%9F%A5%E7%9C%8B%E6%8F%90%E4%BA%A4%E5%8E%86%E5%8F%B2)。

Reviewer 3: 那么多命令，为啥唯独 log 给个 reference？

Reviewer 3: 不要随便 resolve 啊（

???+note
    本页面将着重介绍 Git 这一版本控制系统，与 GitHub 相关的内容，请参考 [GitHub 帮助](https://docs.github.com/cn) 和 [如何参与 - OI Wiki](../intro/htc.md) 。

Git 是目前使用最广泛的版本控制系统之一。 **OI Wiki** 也使用了 Git 作为版本控制系统。

## 安装

### Windows

???+warning
    本节以 Git for Windows 2.28.0（2020-07-27 更新）为例。安装页面的选项可能会随着时间而变动。

Windows 系统可以在 [Git 官网](https://git-scm.com/downloads) 获取到 Git for Windows（下文简称 Git）的安装包。

下载完成后，运行 Git 的安装程序。弹出的页面如下图所示。

![Git for Windows installation 1](images/git3.png)

> 点击 Next >，选择安装路径。

Reviewer: ```suggestion
点击 `Next >`，选择安装路径。
```
下同

```bash
$ git clone https://github.com/OI-wiki/OI-wiki
```

这样在当前文件夹下，就会新建一个名为 `OI-wiki` 的仓库，里面存放着被克隆的仓库的内容。

### 跟踪文件

在对仓库的文件做出了一些修改后，这些修改需要被纳入到版本管理当中去。

使用 `git status` 命令可以查看当前仓库文件的状态。

举个例子，在一个空仓库中新增了一个 `README.md` 文件后，执行 `git status` 命令的效果如下：

```bash
$ git status
On branch master

No commits yet

Untracked files:
  (use "git add <file>..." to include in what will be committed)

        README.md

nothing added to commit but untracked files present (use "git add" to track)
```

这里的 Untracked files 指的是 Git 之前没有纳入版本跟踪的文件。如果文件没有纳入版本跟踪，对该文件的修改不会被 Git 记录。

执行 `git add <文件>` 命令可以将指定的文件纳入到版本跟踪中。

```bash
$ git add README.md # 将这个文件纳入到版本跟踪中
$ git status
On branch master

No commits yet

Changes to be committed:
  (use "git rm --cached <file>..." to unstage)

        new file:   README.md
```

这时 `README.md` 已经纳入了版本跟踪，放入了暂存区。接下来只需执行 `git commit` 命令就可以提交这次更改了。

但在进行这一工作之前，先对 `README.md` 做点小修改。

```bash
$ vim README.md # 随便修改点东西
$ git status
On branch master

No commits yet

Changes to be committed:
  (use "git rm --cached <file>..." to unstage)

        new file:   README.md

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)

>   (use "git checkout -- <file>..." to discard changes in working directory)

Reviewer 3: ```suggestion
  (use "git restore -- <file>..." to discard changes in working directory)
```

[2.28.0](https://github.com/OI-wiki/OI-wiki/pull/2627/files#diff-922cd4d7d078fb7f0b61a02c862b1346R11) 的话应该是 restore。

刚开始自述文件是这样的：

```markdown
# This is a test repo.

This repo includes some c++ codes.
```

在 `readme-refactor` 分支下的自述文件是这样的：

```markdown
# Code Library

This repo includes some c++ codes.
```

在 `master` 分支下的自述文件是这样的：

```markdown
# This is a code library.

This repo includes some c++ codes.
```

这时候运行 `git merge readme-refactor` 命令，Git 提示出现了合并冲突。

执行一下 `git status` 命令，可以查看是哪些文件引发了冲突。

```bash
$ git status
On branch master
You have unmerged paths.
  (fix conflicts and run "git commit")

Unmerged paths:
  (use "git add <file>..." to mark resolution)

    both modified:      README.md

no changes added to commit (use "git add" and/or "git commit -a")
```

如何解决冲突？对于每个发生了合并冲突的文件，Git 都会在这些文件中加入标准的冲突解决标记。比如这个例子中的 `README.md` 文件，打开后它长这个样子：

```markdown
<<<<<< HEAD
# This is a code library.
======
# Code Library
>>>>>> readme-refactor

This repo includes some c++ codes.
```

 `======` 作为分界线将两个分支的内容隔开， `<<<<<< HEAD` 标记和 `======` 之间的部分是 `master` 分支的内容，而 `======` 和 `>>>>>> readme-refactor` 标记之间的部分是 `readme-refactor` 分支的内容。

删除这些冲突解决标记，保存文件，将这些文件纳入暂存区后提交，就可以解决合并冲突了。

```bash
$ git add README.md # 将发生冲突的文件纳入暂存区

> $ git commit -m "merge branch readme-refactor into master"
> [master fe92c6b] merge branch readme-refactor into master

Reviewer 3: ```suggestion
$ git commit -m "Merge branch readme-refactor into master"
[master fe92c6b] Merge branch readme-refactor into master
```

我是觉得不 `-m` 比较好..默认的 merge 标题应该就差不多长这样。

Reviewer 2: > Commit message options 
> -m, --message <message>                   commit message
`-m` 就是一个普通的 commit 命令吧，为什么不要用🤔

Reviewer 3: > 为什么不要用🤔

> 默认的 merge 标题应该就差不多长这样。

Reviewer 3: 就是说，你 `git commit`，文本编辑器里的内容差不多就是 "Merge branch readme-refactor into master"，不需要手动输入。

Author: 嗯，有点道理（

```

## 远程仓库的管理

在本地完成修改后，你可能会需要将这些修改推送到 GitHub 等 Git 仓库托管平台上。托管在这些平台上的仓库就归属于远程仓库的范畴——你可以从这些仓库中获取信息，也可以将你作出的更改推送到远程仓库上。与其他人的协作往往离不开远程仓库，因此学会管理远程仓库很有必要。

### 远程仓库的查看

使用 `git remote` 命令可以查看当前仓库的远程仓库列表。

如果当前仓库是克隆来的，那么应该会有一个叫做 origin 的远程仓库。

```bash
$ git remote
origin
```

### 远程仓库的添加/删除/重命名

执行 `git remote add <name> <url>` 命令可以添加一个名字为 `name` ，链接为 `url` 的远程仓库。

执行 `git remote remame <oldname> <newname>` 可以将名字为 `oldname` 的远程仓库改名为 `newname` 。

执行 `git remote rm <remote>` 可以删除名字为 `name` 的远程仓库。

> 

Reviewer 3: `set-url` 和 `get-url` 也挺有用的，感觉加上比较好。

### 从远程仓库拉取更改

在远程仓库中，其他人可能会推送一些更改，执行 `git fetch` 命令可以将这些更改拉取到本地。

```bash
$ git fetch origin # 拉取 origin 的数据
```

需要注意的是， `git fetch` 命令只会拉取远程仓库的信息，而不会将这些信息合并到本地仓库中。如果需要将这些更改进行合并，需要使用 `git pull` 命令。

```bash
$ git pull origin master # 抓取 origin 的数据并自动和本地的 master 分支合并
```

### 将更改推送到远程仓库

当你完成了一些更改之后，使用 `git push` 命令可以将这些更改推送到远程仓库。

```bash
$ git push origin master # 将 master 分支的数据推送至 origin
```

> 需要注意的是，你的更改能成功推送，需要满足两个条件：你拥有向这个仓库的写入权限，且之前没有人推送过。当远程分支有新更改而当前分支没有时，需要执行 `git pull` 命令完成合并再提交。

Reviewer 3: ```suggestion
需要注意的是，你的更改能成功推送，需要满足两个条件：你拥有向这个仓库（分支）的写入权限，且你的这个分支比远程仓库的这个分支新（可以理解为没有人在你进行修改的这段时间进行了推送）。当远程分支有当前分支没有的新更改时，可以执行 `git pull` 命令完成合并再提交。
```

## 外部链接

-  [Git Reference](https://git-scm.com/docs) 
-  [Pro Git Book](https://git-scm.com/book/zh/v2) 
-  [Learn Git Branching](https://learngitbranching.js.org/) 

## 参考资料与注释

[^note1]: 在某些地方（比如 [LFS 官网](https://git-lfs.github.com/) ）又被称作 Git Large File Storage（大文件存储）。它在将项目托管到平台上时，用文本指针代替音频、视频、图像、数据集等大文件的原始文件，从而加快传输速度。对移动应用程序开发人员、游戏工程师以及任何需要大文件构建软件的人，该功能都极为实用。若想进一步了解该功能，可以参考 [Atlassian 官方介绍 Git LFS 的译文](https://www.cnblogs.com/cangqinglang/p/13097777.html) 。

> [^note2]: 但是，Git for Windows 对 Vim 的描述是“虽然强大，但是可能会难以使用。用户界面反人类，键位映射卡手。Git 使用 Vim 作为默认编辑器只是出于历史原因，强烈推荐换用一个 UI 设计现代化的编辑器。”，并给“难以使用”加上了 [“StackOverflow 每年帮助一百万名开发者退出 Vim”的页面链接](https://stackoverflow.blog/2017/05/23/stack-overflow-helping-one-million-developers-exit-vim/) 。

Reviewer: ```suggestion
[^note2]: 但是，Git for Windows 对 Vim 的描述是“虽然强大，但是可能会难以使用。用户界面反人类，键位映射卡手。Git 使用 Vim 作为默认编辑器只是出于历史原因，强烈推荐换用一个 UI 设计现代化的编辑器。”，并给“难以使用”加上了 [Stack Overflow 每年帮助一百万名开发者退出 Vim](https://stackoverflow.blog/2017/05/23/stack-overflow-helping-one-million-developers-exit-vim/)的页面链接。
```

![Git for Windows installation 4](images/git6.png)

点击 Next >，选择默认的编辑器。

![Git for Windows installation 5](images/git7.png)

在下拉菜单栏中，Git 提供了 nano、 [Vim](editor/vim.md) 、 [Notepad++](editor/npp.md) 、 [Visual Studio Code](editor/vscode.md) 、Visual Studio Code Insider（VSCode 预览版）、Sublime Text、 [Atom](editor/atom.md) 和 VSCodium 作为默认编辑器的预设方案。如果系统已经安装了上述的一种或多种编辑器，可以直接选择其中一种并进入下一步。另外，还可通过 `Select other editor as Git's default editor` 项手动指定默认编辑器。

![Git for Windows installation 6](images/git8.png)

除了 Vim 外，如果选择了尚未安装的编辑器，安装程序会阻止使用者进入下一步。[^note2]用户可以通过给出的官网链接跳转到下载页面，完成编辑器的安装后继续安装 Git。

![Git for Windows installation 7](images/git9.png)

选好默认编辑器后，点击 Next > 进入下一步。

接下来的每一步都直接点击 Next >。当 Next > 变为 Install 时，按下 Install 进行安装。

最后，该页面出现，Git 安装完成。此时可以取消勾选 `View Release Notes` ，然后按下 Next > 直接退出；也可以直接按下 Next > 查看版本更新信息；亦可以勾选 `Launch Git Bash` 运行 Git 命令行。

![Git for Windows installation 8](images/git10.png)

### Linux

Linux 系统可以通过包管理器进行安装。不同 Linux 发行版的安装方式可能不同， [Git 官网](https://git-scm.com/download/linux) 给出了不同系统下的安装方式。

### macOS

macOS 系统只需在终端执行 `brew install git` 命令即可安装。（需安装 [Homebrew](https://brew.sh/) ）

## 配置

Git 的配置文件分为三个级别：

1. 对系统上的所有用户，以及该系统上的所有仓库有效的配置文件。

> 2. 对系统的当前用户，以及该系统上的所有仓库有效的配置文件。

Reviewer 3: ```suggestion
2. 对系统的当前用户，以及该用户的所有仓库有效的配置文件。
```

不知道有没有更好的表达方式 :thinking: “该用户操作的所有仓库”？

???+note
    本页面将着重介绍 Git 这一版本控制系统，与 GitHub 相关的内容，请参考 [GitHub 帮助](https://docs.github.com/cn) 和 [如何参与 - OI Wiki](../intro/htc.md) 。

Git 是目前使用最广泛的版本控制系统之一。 **OI Wiki** 也使用了 Git 作为版本控制系统。

## 安装

### Windows

???+warning
    本节以 Git for Windows 2.28.0（2020-07-27 更新）为例。安装页面的选项可能会随着时间而变动。

Windows 系统可以在 [Git 官网](https://git-scm.com/downloads) 获取到 Git for Windows（下文简称 Git）的安装包。

下载完成后，运行 Git 的安装程序。弹出的页面如下图所示。

![Git for Windows installation 1](images/git3.png)

点击 `Next >` ，选择安装路径。

![Git for Windows installation 2](images/git4.png)

> 点击 Next >，选择安装可选项。

Reviewer 5: 19行的next有 ` `` ` 包裹, 这一行的没2333
是否需要统一?

Reviewer 5: 下面的next好像也都没有

### macOS

macOS 系统只需在终端执行 `brew install git` 命令即可安装。（需安装 [Homebrew](https://brew.sh/) ）

## 配置

Git 根据配置文件的应用范围，将配置文件分为不同的等级，其中较常用的有两个级别[^note3]：

1. 适用于当前用户的全局配置文件，该用户操作本系统上的所有仓库时都会查询该配置文件。
2. 适用于当前仓库的配置文件。

当多个配置文件对同一个选项作出设置的时候，局部设置会自动覆盖全局设置。因此如果需要在某个仓库应用特定的设置的话，只需更改该仓库下的特定设置即可，不会对全局设置造成影响。

修改配置文件需要用到 `git config` 命令。

### 设置用户信息

安装 Git 后，第一件事情就是设置你的用户名和邮箱。这些信息在每次提交时都会用到。

```bash
$ git config --global user.name "OI Wiki"
$ git config --global user.email oi-wiki@example.com
```

???+note
    这里给出的用户名和邮箱仅供演示。您在根据本页面的内容配置时，请记得将这里的用户名和邮箱改成自己的信息。

这里的 `--global` 表示修改的是全局配置，即该设置对当前用户下的所有仓库均有效。如果不添加 `--global` 选项，则会默认修改当前仓库下的配置文件。

如果想要修改某个仓库的特定设置，只需在该仓库下执行不带 `--global` 的命令即可。

### 配置编辑器

```bash
$ git config --global core.editor emacs
```

执行如上命令可以将编辑器更改为 [Emacs](./editor/emacs.md) 。

在 Windows 下，Git 的默认编辑器可以在安装 Git 时选择。之后若要修改，在 Git Bash 里输入如上命令，将编辑器名换成编辑器的绝对路径，运行命令即可。

### 显示配置

可以通过 `git config -l` 列出当前已经设置的所有配置参数。使用 `git config --global -l` 可以列出所有全局配置。

## 仓库操作基础

### 新建 Git 仓库

新建一个 Git 仓库非常简单，只需在想要建立仓库的文件夹输入如下命令：

```bash
$ git init
```

Git 将在当前文件夹新建一个 `.git` 文件夹，一个仓库就这样建好了。

如果想把一个仓库拷贝到自己的电脑上（比如将 **OI Wiki** 的代码拷贝到本地上进行编辑），采用 `git clone` 命令即可。

```bash

> $ git clone https://github.com/OI-wiki/OI-wiki

Reviewer 6: 要不要加个 ssh 的链接？

Reviewer 7: 我觉得是比较有必要的，GitHub 把用户名+密码的连接方式干掉了，现在只能用户名+Personal Token或者 SSH，其中 SSH 相对简单和方便些

```

这样在当前文件夹下，就会新建一个名为 `OI-wiki` 的仓库，里面存放着被克隆的仓库的内容。

### 跟踪文件

在对仓库的文件做出了一些修改后，这些修改需要被纳入到版本管理当中去。

使用 `git status` 命令可以查看当前仓库文件的状态。

举个例子，在一个空仓库中新增了一个 `README.md` 文件后，执行 `git status` 命令的效果如下：

```bash
$ git status
On branch master

No commits yet

Untracked files:
  (use "git add <file>..." to include in what will be committed)

        README.md

nothing added to commit but untracked files present (use "git add" to track)
```

这里的 Untracked files 指的是 Git 之前没有纳入版本跟踪的文件。如果文件没有纳入版本跟踪，对该文件的修改不会被 Git 记录。

> 执行 `git add <文件>` 命令可以将指定的文件纳入到版本跟踪中。

Reviewer 8: 这个“文件”是否有需要说明是相对当前命令行路径

Author: 个人认为没有太大必要 :thinking:，毕竟命令引用的路径，大都是相对当前的路径。

> 执行 `git add <文件>` 命令可以将指定的文件纳入到版本跟踪中。

Reviewer 6: 要不要提及一下 `git add .`？

Author: > 要不要提及一下 `git add .`？

有 `git add -A` 了（

执行一下 `git status` 命令，可以查看是哪些文件引发了冲突。

```bash
$ git status
On branch master
You have unmerged paths.
  (fix conflicts and run "git commit")

Unmerged paths:
  (use "git add <file>..." to mark resolution)

    both modified:      README.md

no changes added to commit (use "git add" and/or "git commit -a")
```

如何解决冲突？对于每个发生了合并冲突的文件，Git 都会在这些文件中加入标准的冲突解决标记。比如这个例子中的 `README.md` 文件，打开后它长这个样子：

```markdown
<<<<<< HEAD
# This is a code library.
======
# Code Library
>>>>>> readme-refactor

This repo includes some c++ codes.
```

 `======` 作为分界线将两个分支的内容隔开， `<<<<<< HEAD` 标记和 `======` 之间的部分是 `master` 分支的内容，而 `======` 和 `>>>>>> readme-refactor` 标记之间的部分是 `readme-refactor` 分支的内容。

删除这些冲突解决标记，保存文件，将这些文件纳入暂存区后提交，就可以解决合并冲突了。

```bash
$ git add README.md # 将发生冲突的文件纳入暂存区
$ git commit
[master fe92c6b] Merge branch readme-refactor into master
```

## 管理远程仓库

在本地完成修改后，你可能会需要将这些修改推送到 GitHub 等 Git 仓库托管平台上。托管在这些平台上的仓库就归属于远程仓库的范畴——你可以从这些仓库中获取信息，也可以将你作出的更改推送到远程仓库上。与其他人的协作往往离不开远程仓库，因此学会管理远程仓库很有必要。

### 远程仓库的查看

使用 `git remote` 命令可以查看当前仓库的远程仓库列表。

如果当前仓库是克隆来的，那么应该会有一个叫做 origin 的远程仓库。

```bash
$ git remote
origin
```

### 远程仓库的配置

执行 `git remote add <name> <url>` 命令可以添加一个名字为 `name` ，链接为 `url` 的远程仓库。

> 执行 `git remote remame <oldname> <newname>` 可以将名字为 `oldname` 的远程仓库改名为 `newname` 。

Reviewer 8: ```suggestion
执行 `git remote rename <oldname> <newname>` 可以将名字为 `oldname` 的远程仓库改名为 `newname` 。
```

另外这里可能要说明一下不选 oldname 和 newname 时的行为

???+note
    本页面将着重介绍 Git 这一版本控制系统，与 GitHub 相关的内容，请参考 [GitHub 帮助](https://docs.github.com/cn) 和 [如何参与 - OI Wiki](../intro/htc.md)。

Git 是目前使用最广泛的版本控制系统之一。**OI Wiki** 也使用了 Git 作为版本控制系统。

## 安装

### Windows

???+warning
    本节以 Git for Windows 2.28.0（2020-07-27 更新）为例。安装页面的选项可能会随着时间而变动。

Windows 系统可以在 [Git 官网](https://git-scm.com/downloads) 获取到 Git for Windows（下文简称 Git）的安装包。

下载完成后，运行 Git 的安装程序。弹出的页面如下图所示。

![Git for Windows installation 1](images/git3.png)

点击 `Next >`，选择安装路径。

![Git for Windows installation 2](images/git4.png)

点击 `Next >`，选择安装可选项。

![Git for Windows installation 3](images/git5.png)

默认的选项如上图所示，均不建议取消勾选。其他未勾选的选项可以依照需求勾选。

??? note "各选项功能详解"
    默认勾选的选项：
    
    -   `Windows Explorer integration`：资源管理器集成（右键菜单）。
        - `Git Bash Here`：在当前目录运行 Git 命令行。
        - `Git GUI Here`：在当前目录运行 Git 图形化页面。
    - `Git LFS`：安装 Git Large File Support（大文件支持）。[^note1]
    - `Associate .git* configuration files with the default text editor`：将所有 `.git` 配置文件与默认文本编辑器关联，并用其打开。默认文本编辑器会在后面配置。

>     - `Associate .sh files to be run with Bash`：将所有扩展名为 `.sh` 的文件与 Git 命令行关联，并用其执行。

Reviewer 8: ```suggestion
    - `Associate .sh files to be run with Bash`：将所有扩展名为 `.sh` 的文件与 Git Bash 关联，并用其执行。
```

在下拉菜单栏中，Git 提供了 nano、[Vim](editor/vim.md)、[Notepad++](editor/npp.md)、[Visual Studio Code](editor/vscode.md)、Visual Studio Code Insiders（VS Code 预览版）、Sublime Text、[Atom](editor/atom.md) 和 VSCodium 作为默认编辑器的预设方案。如果系统已经安装了上述的一种或多种编辑器，可以直接选择其中一种并进入下一步。另外，还可通过 `Select other editor as Git's default editor` 项手动指定默认编辑器。

![Git for Windows installation 6](images/git8.png)

除了 Vim 外，如果选择了尚未安装的编辑器，安装程序会阻止使用者进入下一步。[^note2]用户可以通过给出的官网链接跳转到下载页面，完成编辑器的安装后继续安装 Git。

![Git for Windows installation 7](images/git9.png)

选好默认编辑器后，点击 `Next >` 进入下一步。

接下来的每一步都直接点击 `Next >`。当 `Next >` 变为 `Install` 时，按下 `Install` 进行安装。

最后，该页面出现，Git 安装完成。此时可以取消勾选 `View Release Notes`，然后按下 `Next >` 直接退出；也可以直接按下 `Next >` 查看版本更新信息；亦可以勾选 `Launch Git Bash` 运行 Git 命令行。

![Git for Windows installation 8](images/git10.png)

### Linux

Linux 系统可以通过包管理器进行安装。不同 Linux 发行版的安装方式可能不同，[Git 官网](https://git-scm.com/download/linux) 给出了不同系统下的安装方式。

### macOS

macOS 系统只需在终端执行 `brew install git` 命令即可安装。（需安装 [Homebrew](https://brew.sh/)）

## 配置

Git 根据配置文件的应用范围，将配置文件分为不同的等级，其中较常用的有两个级别[^note3]：

1. 适用于当前用户的全局配置文件，该用户操作本系统上的所有仓库时都会查询该配置文件。
2. 适用于当前仓库的配置文件。

当多个配置文件对同一个选项作出设置的时候，局部设置会自动覆盖全局设置。因此如果需要在某个仓库应用特定的设置的话，只需更改该仓库下的特定设置即可，不会对全局设置造成影响。

> 修改配置文件需要用到 `git config` 命令。

Reviewer 8: ```suggestion
修改配置文件需要用到 `git config` 命令，加上 `--global` 开关后则写入适用于当前用户的全局设置。
```

Author: 其实这个在下一节已经提及过了（

### 设置用户信息

安装 Git 后，第一件事情就是设置你的用户名和邮箱。这些信息在每次提交时都会用到。

```bash
$ git config --global user.name "OI Wiki"
$ git config --global user.email oi-wiki@example.com
```

???+note
    这里给出的用户名和邮箱仅供演示。您在根据本页面的内容配置时，请记得将这里的用户名和邮箱改成自己的信息。

这里的 `--global` 表示修改的是全局配置，即该设置对当前用户下的所有仓库均有效。如果不添加 `--global` 选项，则会默认修改当前仓库下的配置文件。

如果想要修改某个仓库的特定设置，只需在该仓库下执行不带 `--global` 的命令即可。

### 配置编辑器

```bash
$ git config --global core.editor emacs
```

执行如上命令可以将编辑器更改为 [Emacs](./editor/emacs.md)。

在 Windows 下，Git 的默认编辑器可以在安装 Git 时选择（见前文）。之后若要修改，在 Git Bash 里输入如上命令，将编辑器名换成编辑器的绝对路径，运行命令即可。

### 显示配置

可以通过 `git config -l` 列出当前已经设置的所有配置参数。使用 `git config --global -l` 可以列出所有全局配置。

## 仓库操作基础

### 新建 Git 仓库

新建一个 Git 仓库非常简单，只需在想要建立仓库的文件夹输入如下命令：

```bash
$ git init
```

Git 将在当前文件夹新建一个 `.git` 文件夹，一个仓库就这样建好了。

如果想把一个仓库拷贝到自己的电脑上（比如将 **OI Wiki** 的代码拷贝到本地上进行编辑），采用 `git clone` 命令即可。

```bash
$ git clone https://github.com/OI-wiki/OI-wiki
```

> 这样在当前文件夹下，就会新建一个名为 `OI-wiki` 的仓库，里面存放着被克隆的仓库的内容。

Reviewer 8: ```suggestion
这样，被克隆的仓库的内容就会被储存到当前文件夹下一个与仓库同名的新文件夹。在本例中，当前文件夹下会出现一个名为 `OI-wiki` 的新文件夹。
```

### 跟踪文件

在对仓库的文件做出了一些修改后，这些修改需要被纳入到版本管理当中去。

使用 `git status` 命令可以查看当前仓库文件的状态。

举个例子，在一个空仓库中新增了一个 `README.md` 文件后，执行 `git status` 命令的效果如下：

```bash
$ git status
On branch master

No commits yet

Untracked files:
  (use "git add <file>..." to include in what will be committed)

        README.md

nothing added to commit but untracked files present (use "git add" to track)
```

这里的 Untracked files 指的是 Git 之前没有纳入版本跟踪的文件。如果文件没有纳入版本跟踪，对该文件的修改不会被 Git 记录。

执行 `git add <文件>` 命令可以将指定的文件纳入到版本跟踪中。

```bash
$ git add README.md # 将这个文件纳入到版本跟踪中
$ git status
On branch master

No commits yet

Changes to be committed:
  (use "git rm --cached <file>..." to unstage)

        new file:   README.md
```

这时 `README.md` 已经纳入了版本跟踪，放入了暂存区。接下来只需执行 `git commit` 命令就可以提交这次更改了。

> 但在进行这一工作之前，先对 `README.md` 做点小修改。

Reviewer 8: ```suggestion
但在进行这一工作之前，让我们先对 `README.md` 做点小修改。
```

表示不是必须的？

```bash
$ vim README.md # 随便修改点东西
$ git status
On branch master

No commits yet

Changes to be committed:
  (use "git rm --cached <file>..." to unstage)

        new file:   README.md

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore -- <file>..." to discard changes in working directory)

        modified:   README.md
```

> 你会发现 `README.md` 同时处于暂存区和非暂存区。如果这时候执行 `git commit` 命令，只有处于暂存区的更改会被提交，而非暂存区的修改，则不会被提交。

Reviewer 3: ```suggestion
你会发现 `README.md` 同时处于暂存区和非暂存区。实际上，是否处于暂存区是对于更改而言的，而不是对于文件而言的，所以对 `README.md` 的前一次修改已被纳入暂存区，而后一次修改还没有。如果这时候执行 `git commit` 命令，只有处于暂存区的更改会被提交，而非暂存区的更改，则不会被提交。
```

Git 给了一条提示，执行 `git add README.md` 就可以将非暂存区的更改放入暂存区了。

???+note "一次性将所有更改放入暂存区"
    `git add` 命令会将对指定的文件的修改放入暂存区中。
    
    在多数情况下，用户更期望一次性将所有更改都放入暂存区中，这时候可以应用 `git add -A` 命令。该命令会将所有更改（包括未被纳入版本跟踪的文件，不包括被忽略的文件）放入暂存区。

> 

Reviewer 3: ```suggestion

    如果只需更新已被纳入版本跟踪的文件，而不将未纳入版本跟踪的文件加入暂存区，可以使用 `git add -u`。

```

现在将非暂存区的文件加入暂存区，将所有更改一并提交。

```bash
$ git add README.md
$ git commit # 接下来会弹出编辑器页面，你需要写下 commit 信息
[master (root-commit) f992763] initial commit
 1 file changed, 2 insertions(+)
 create mode 100644 README.md
```

现在重点观察一下 commit 的信息。

`master` 表示当前位于 `master` 分支（关于分支的问题，下文将会详细介绍），`b13c84e` 表示本次提交的 SHA-1 校验和的前几位，后面则是本次提交的信息。

需要特别关注的是这里的 SHA-1 校验码，每个校验码都与某个时刻仓库的一个快照相对应。利用这一特性我们可以访问历史某个时刻的仓库快照，并在该快照上进行更改。

接下来两行则详细说明了本次更新涉及的文件修改。

另外，commit 过程中可以利用几个参数来简化提交过程：

- `-a`：在提交前将所有已跟踪的文件的更改放入暂存区。需要注意的是未被跟踪的文件（新创建的文件）不会被自动加入暂存区，需要用 `git add` 命令手动添加。

> - `-m`：该参数后跟提交信息，表示以该提交信息提交本次更改。

Reviewer 8: ```suggestion
- `-m`：该参数后跟提交信息，表示以该提交信息提交本次更改。例如 `git commit -m "fix: typo` 会创建一条标题为 `fix: typo` 的 commit。
```

[master 5ca15f0] feat: update README.md
 1 file changed, 1 insertion(+), 1 deletion(-)
```

下面用一张图来解释刚才的操作过程。

![](./images/git1.png)

`master` 分支被标红，表明在这几次操作后，它是当前分支（即 HEAD 指向的位置）。

- 最开始时 `master` 指向 `ae9dd37` 这一快照。
- 接下来在 `master` 所在的位置创建了一个新的 dev 分支，该分支一开始和 master 指向相同位置。
- 在 `dev` 分支上作了一些修改（创建了 `aplusb.cpp`），进行了一次提交，本次提交后，`dev` 分支指向 `5da093b` 这一快照。
- 切换回 `master` 分支后，因为 `master` 分支还指向 `ae9dd37`，还没有创建 `aplusb.cpp`，因此仓库中没有这一文件。
- 接下来在 `master` 分支上进行修改（更新了 `README.md`），进行了一次提交，`master` 分支指向了 `5ca15f0` 这一快照。

### 分支的合并

当一个分支上的工作已经完成，就可以将这些工作合并到另外一个分支上去。

还是接着上面这个例子，`dev` 分支的工作已经完成，通过 `git merge` 命令可以将该分支合并到当前分支（`master`）上：

```bash
$ git merge dev
Merge made by the 'recursive' strategy.
 aplusb.cpp | 7 +++++++
 1 file changed, 7 insertions(+)
 create mode 100644 aplusb.cpp
```

![](./images/git2.png)

这次合并具体是怎么执行的呢？

在合并之前，`master` 指向 `5ca15f0`，而 `dev` 指向 `5da093b`，这两个状态并不在一条链上。

Git 会找到这两个状态的最近公共祖先（在上图中是 `ae9dd37`），并对这三个快照进行一次合并。三个快照合并的结果作为一个新的快照，并将当前分支指向这一快照。

合并过程本身也是一次提交，不过与常规提交不同的是，合并提交有不止一个前驱提交，它是多个提交状态合并后的结果。

在合并完成后，`dev` 分支就完成了它的使命，这时候可以利用下面的命令删除 `dev` 分支：

```bash
$ git branch -d dev
```

不过合并过程并非总是这么顺利，在某些情况下，合并过程可能会出现冲突，这个问题接下来会讲到。

### 解决合并冲突

> 如果在两个分支中，对同一个文件的同一部分进行了不同的修改，Git 就无法自动合并这两个分支，也就是发生了合并冲突。

Reviewer 8: ```suggestion
如果在两个分支中，对同一个文件的同一行号进行了不同的修改，Git 就无法自动合并这两个分支，也就是发生了合并冲突。
```

Reviewer 4: 这个是行号吗 🤔 ，我觉得还是改成部分吧？

Reviewer 4: 和diff的算法相关，我感觉

接着上面的例子，假如你在合并后的 `master` 分支的基础上，新开了一个 `readme-refactor` 分支，准备重写一份自述文件。但因为一些疏忽，你同时修改了 `readme-refactor` 和 `master` 分支的自述文件。

刚开始自述文件是这样的：

```markdown
# This is a test repo.

This repo includes some c++ codes.
```

在 `readme-refactor` 分支下的自述文件是这样的：

```markdown
# Code Library

This repo includes some c++ codes.
```

在 `master` 分支下的自述文件是这样的：

```markdown
# This is a code library.

This repo includes some c++ codes.
```

这时候运行 `git merge readme-refactor` 命令，Git 提示出现了合并冲突。

执行一下 `git status` 命令，可以查看是哪些文件引发了冲突。

```bash
$ git status
On branch master
You have unmerged paths.
  (fix conflicts and run "git commit")

Unmerged paths:
  (use "git add <file>..." to mark resolution)

    both modified:      README.md

no changes added to commit (use "git add" and/or "git commit -a")
```

如何解决冲突？对于每个发生了合并冲突的文件，Git 都会在这些文件中加入标准的冲突解决标记。比如这个例子中的 `README.md` 文件，打开后它长这个样子：

```markdown
<<<<<< HEAD
# This is a code library.
======
# Code Library
>>>>>> readme-refactor

This repo includes some c++ codes.
```

> `======` 作为分界线将两个分支的内容隔开，`<<<<<< HEAD` 标记和 `======` 之间的部分是 `master` 分支的内容，而 `======` 和 `>>>>>> readme-refactor` 标记之间的部分是 `readme-refactor` 分支的内容。

Reviewer 3: ```suggestion
`======` 作为分界线将两个分支的内容隔开，`<<<<<< HEAD` 标记和 `======` 之间的部分是 HEAD 指针（`master` 分支）的内容，而 `======` 和 `>>>>>> readme-refactor` 标记之间的部分是 `readme-refactor` 分支的内容。
```

> 删除这些冲突解决标记，保存文件，将这些文件纳入暂存区后提交，就可以解决合并冲突了。

Reviewer 3: ```suggestion
解决冲突，删除这些冲突标记，保存文件，将这些文件纳入暂存区后提交，就可以解决合并冲突了。
```

```bash
$ git add README.md # 将发生冲突的文件纳入暂存区
$ git commit
[master fe92c6b] Merge branch readme-refactor into master
```

### 其他合并方式

默认情况下，Git 采用 Merge（合并）的方式合并两个分支。使用该方法将分支 B 并入分支 A 时，会将 B 分支的所有 commit 并入 A 分支的提交历史中。

除此以外，Git 还提供了两种合并分支的方式：Squash（压缩）和 Rebase（变基）。

#### Squash（压缩）

使用 Squash 方式将分支 B 并入分支 A 时，在 B 分支上的所有更改会被合并为一次 commit 提交到 A 分支。

在 `git merge` 中加入 `--squash` 参数即可使用 Squash 方式进行分支合并。

```bash
git merge <branch> --squash
```

需要注意的是，在执行上述命令后，Git 只会将 B 分支的所有更改存入 A 分支的缓冲区内，接下来还需要执行一次 `git commit` 命令完成合并工作。

> 使用 Squash 方式合并可以简化 commit 记录，但是会丢失一些原分支上的 commit 信息（包括原分支上每次 commit 的提交者，每次 commit 的具体更改等等）。

Reviewer 3: ```suggestion
使用 Squash 方式合并可以简化 commit 记录，但是会丢失具体到每一次 commit 的信息（每次 commit 的提交者，每次 commit 的更改等等），只留下合并为一个整体的信息（每次 commit 的提交者会以 "Co-authored-by" 的形式在提交信息中列出）。

但如果是在 GitHub 上进行 Squash and Merge，原有的信息都可以在 Pull Request 中查看。
```

#### Rebase（变基）

使用 Rebase 方式将分支 B 并入分支 A 时，在 B 分支上的每一次 commit 都会单独添加到 A 分支，而不再像 Merge 方式那样创建一个合并 commit 来合并两个分支的内容[^note4]。

首先，切换到 B 分支，接下来将 B 分支变基到 A 分支：

```bash
git checkout B
git rebase A
```

现在切回到 A 分支，再执行一次 `git merge` 命令，即可完成将 B 分支的内容合并到 A 分支的工作。

```bash
git checkout A
git merge B
```

> 使用 Rebase 完成合并可以让 commit 历史线性化。但是这样做的缺点也很明显：Rebase 会改变已有的 commit 历史，在多人协作时会有出现冲突或丢失更改的风险。

Reviewer 8: 风险例如？

Reviewer 4: 应该是指不要rebase在远程的分支？

Reviewer 8: 那也不对啊 🤔 
但是 rebase 确实是对你本地修改有风险倒是

Reviewer 4: ![image](https://user-images.githubusercontent.com/25521218/121877696-78e67380-cd3d-11eb-954c-ca60ed9910cb.png)

Reviewer 3: > ![image](https://user-images.githubusercontent.com/25521218/121877696-78e67380-cd3d-11eb-954c-ca60ed9910cb.png)

这个“如果提交存在于你的仓库之外”的“提交”指代不明，TA 想表达的应该是“被 rebase 改变的 commits”。

Reviewer 8: > > ![image](https://user-images.githubusercontent.com/25521218/121877696-78e67380-cd3d-11eb-954c-ca60ed9910cb.png)
> 
> 这个“如果提交存在于你的仓库之外”的“提交”指代不明，TA 想表达的应该是“被 rebase 改变的 commits”。

@ouuan  倒是 get 到了（

> 使用 Rebase 完成合并可以让 commit 历史线性化。但是这样做的缺点也很明显：Rebase 会改变已有的 commit 历史，在多人协作时会有出现冲突或丢失更改的风险。

Reviewer 3: ```suggestion
使用 Rebase 完成合并可以让 commit 历史线性化，在适当的场景下正确地使用 Rebase 可以达到比 Merge 更好的效果。但是这样做会改变提交历史，在进行 Rebase 时和 Rebase 后再进行相关合并操作时都会增加出现冲突的可能，如果操作不当可能反而会使提交历史变得杂乱。因此，如果对 Rebase 操作没有充分的了解，不建议使用。
```

## 管理远程仓库

在本地完成修改后，你可能会需要将这些修改推送到 GitHub 等 Git 仓库托管平台上。托管在这些平台上的仓库就归属于远程仓库的范畴——你可以从这些仓库中获取信息，也可以将你作出的更改推送到远程仓库上。与其他人的协作往往离不开远程仓库，因此学会管理远程仓库很有必要。

### 远程仓库的查看

使用 `git remote` 命令可以查看当前仓库的远程仓库列表。

如果当前仓库是克隆来的，那么应该会有一个叫做 origin 的远程仓库。

```bash
$ git remote
origin
```

### 远程仓库的配置

执行 `git remote add <name> <url>` 命令可以添加一个名字为 `name`，链接为 `url` 的远程仓库。

执行 `git remote rename <oldname> <newname>` 可以将名字为 `oldname` 的远程仓库改名为 `newname`。

执行 `git remote rm <name>` 可以删除名字为 `name` 的远程仓库。

执行 `git remote get-url <name>` 可以查看名字为 `name` 的远程仓库的链接。

执行 `git remote set-url <name> <newurl>` 可以将名字为 `name` 的远程仓库的链接更改为 `newurl`。

### 从远程仓库拉取更改

> 在远程仓库中，其他人可能会推送一些更改，执行 `git fetch` 命令可以将这些更改拉取到本地。

Reviewer 3: ```suggestion
在远程仓库中，其他人可能会推送一些更改，执行 `git fetch` 命令可以将这些更改获取到本地。
```

“拉取”感觉就很 pull :thinking:

```bash

> $ git fetch <remote-name> # 拉取 <remote-name> 的数据

Reviewer 3: ```suggestion
$ git fetch <remote-name> # 获取 <remote-name> 的更改
```

```

> 需要注意的是，`git fetch` 命令只会拉取远程仓库的信息，而不会将这些信息合并到本地仓库中。如果需要将这些更改进行合并，需要使用 `git pull` 命令。

Reviewer 3: ```suggestion
需要注意的是，`git fetch` 命令只会获取远程仓库的更改，而不会将这些更改合并到本地仓库中。如果需要将这些更改进行合并，可以使用 `git pull` 命令。在默认情况下，`git pull` 相当于 `git fetch` 后 `git merge FETCH_HEAD`。
```

```bash

> $ git pull <remote-name> <branch> # 抓取 <remote-name> 的数据并自动和本地的 <branch> 分支合并

Reviewer 3: ```suggestion
$ git pull <remote-name> <branch> # 获取 <remote-name> 的更改，然后将这些更改合并到 HEAD。
```

```

### 将更改推送到远程仓库

当你完成了一些更改之后，使用 `git push` 命令可以将这些更改推送到远程仓库。

```bash

> $ git push <remote-name> <branch> # 将 <branch> 分支的数据推送至 <remote-name>

Reviewer 3: ```suggestion
$ git push <remote-name> <branch> # 将 <branch> 分支的更改推送至 <remote-name> 的 <branch> 分支
```

```

需要注意的是，你的更改能成功推送，需要满足两个条件：你拥有向这个仓库（分支）的写入权限，且你的这个分支比远程仓库的相应分支新（可以理解为没有人在你进行修改的这段时间进行了推送）。当远程分支有当前分支没有的新更改时，可以执行 `git pull` 命令完成合并再提交。

> 

Reviewer 3: `push -f` 和 `push --force-with-lease` :thinking:

### 追踪远程分支

通过将一个本地分支设定为追踪远程分支，可以方便地查看本地分支与远程分支的差别，并能简化与远程分支交互时的操作。

在开始追踪前，你需要先执行 `git fetch <remote-name>` 将远程仓库的信息抓取到本地。

> 接下来执行 `git switch <remote-name>/<remote-branch>` 切换到你想切换的远程分支。该操作会在本地自动创建名字为 `<remote-branch>` 的新分支，并设定该分支自动追踪相应的远程分支。

Reviewer 3: ```suggestion
接下来执行 `git switch <remote-branch>`，会在本地自动创建名字为 `<remote-branch>` 的新分支，并设定该分支自动追踪相应的远程分支。
```

不能 switch 到远程分支吧，有测试过吗

Author: 看起来并不 work，可能晚上写这这段的时候有点犯困 :sleeping:

下载完成后，运行 Git 的安装程序。弹出的页面如下图所示。

![Git for Windows installation 1](images/git3.png)

点击 `Next >` ，选择安装路径。

![Git for Windows installation 2](images/git4.png)

点击 `Next >`，选择安装可选项。

![Git for Windows installation 3](images/git5.png)

默认的选项如上图所示，均不建议取消勾选。其他未勾选的选项可以依照需求勾选。

??? note "各选项功能详解"

    默认勾选的选项：

    - `Windows Explorer integration` ：资源管理器集成（右键菜单）。
        - `Git Bash Here` ：在当前目录运行 Git 命令行。
        - `Git GUI Here` ：在当前目录运行 Git 图形化页面。
    - `Git LFS` ：安装 Git Large File Support（大文件支持）。[^note1]
    - `Associate .git* configuration files with the default text editor` ：将所有 `.git` 配置文件与默认文本编辑器关联，并用其打开。默认文本编辑器会在后面配置。
    - `Associate .sh files to be run with Bash` ：将所有扩展名为 `.sh` 的文件与 Git 命令行关联，并用其执行。

    ---

    默认不勾选的选项：

    - `Additional icons`
        - `On the Desktop`：让 Git 在桌面创建快捷方式。
    - `Use a TrueType font in all console windows` ：让 Git 在所有命令行（包括 cmd 命令行窗口）中都使用 TrueType 字体。
    - `Check daily for Git for Windows updates`：让 Git 每天检查更新。

点击 `Next >`，选择存放 Git 快捷方式的开始菜单文件夹。若勾选 `Don't create a Start Menu folder` ，则 Git 不会在开始菜单创建快捷方式。

![Git for Windows installation 4](images/git6.png)

点击 `Next >`，选择默认的编辑器。

![Git for Windows installation 5](images/git7.png)

> 在下拉菜单栏中，Git 提供了 nano、 [Vim](editor/vim.md) 、 [Notepad++](editor/npp.md) 、 [Visual Studio Code](editor/vscode.md) 、Visual Studio Code Insider（VSCode 预览版）、Sublime Text、 [Atom](editor/atom.md) 和 VSCodium 作为默认编辑器的预设方案。如果系统已经安装了上述的一种或多种编辑器，可以直接选择其中一种并进入下一步。另外，还可通过 `Select other editor as Git's default editor` 项手动指定默认编辑器。

Reviewer: ```suggestion
在下拉菜单栏中，Git 提供了 nano、 [Vim](editor/vim.md) 、 [Notepad++](editor/npp.md) 、 [Visual Studio Code](editor/vscode.md) 、Visual Studio Code Insider（VS Code 预览版）、Sublime Text、 [Atom](editor/atom.md) 和 VSCodium 作为默认编辑器的预设方案。如果系统已经安装了上述的一种或多种编辑器，可以直接选择其中一种并进入下一步。另外，还可通过 `Select other editor as Git's default editor` 项手动指定默认编辑器。
```

在 `git merge` 中加入 `--squash` 参数即可使用 Squash 方式进行分支合并。

```bash
git merge <branch> --squash
```

需要注意的是，在执行上述命令后，Git 只会将 B 分支的所有更改存入 A 分支的缓冲区内，接下来还需要执行一次 `git commit` 命令完成合并工作。

使用 Squash 方式合并可以简化 commit 记录，但是会丢失具体到每一次 commit 的信息（每次 commit 的提交者，每次 commit 的更改等等），只留下合并为一个整体的信息（每次 commit 的提交者会以 "Co-authored-by" 的形式在提交信息中列出）。但如果是在 GitHub 上进行 Squash and Merge，原有的信息都可以在 Pull Request 中查看。

#### Rebase（变基）

使用 Rebase 方式将分支 B 并入分支 A 时，在 B 分支上的每一次 commit 都会单独添加到 A 分支，而不再像 Merge 方式那样创建一个合并 commit 来合并两个分支的内容[^note4]。

首先，切换到 B 分支，接下来将 B 分支变基到 A 分支：

```bash
git checkout B
git rebase A
```

现在切回到 A 分支，再执行一次 `git merge` 命令，即可完成将 B 分支的内容合并到 A 分支的工作。

```bash
git checkout A
git merge B
```

使用 Rebase 完成合并可以让提交历史线性化，在适当的场景下正确地使用 Rebase 可以达到比 Merge 更好的效果。但是这样做会改变提交历史，在进行 Rebase 时和 Rebase 后再进行相关合并操作时都会增加出现冲突的可能，如果操作不当可能反而会使提交历史变得杂乱。因此，如果对 Rebase 操作没有充分的了解，不建议使用。

## 管理远程仓库

在本地完成更改后，你可能会需要将这些更改推送到 GitHub 等 Git 仓库托管平台上。托管在这些平台上的仓库就归属于远程仓库的范畴——你可以从这些仓库中获取信息，也可以将你作出的更改推送到远程仓库上。与其他人的协作往往离不开远程仓库，因此学会管理远程仓库很有必要。

### 远程仓库的查看

使用 `git remote` 命令可以查看当前仓库的远程仓库列表。

如果当前仓库是克隆来的，那么应该会有一个叫做 origin 的远程仓库，它的链接就是克隆时用的链接。

```bash
$ git remote
origin
```

如果要查看某个远程仓库的详细信息的话，可以这样操作：

```bash

> $ git remote show origin
> * 远程 origin
>   获取地址：git@github.com:OI-wiki/OI-wiki.git
>   推送地址：git@github.com:OI-wiki/OI-wiki.git
>   HEAD 分支：master
>   远程分支：
>     git             已跟踪
>     master          已跟踪
>   ...

Reviewer 6: 应该是英文吧（

```suggestion
$ git remote show origin
* remote origin
  Fetch URL: git@github.com:OI-wiki/OI-wiki.git
  Push  URL: git@github.com:OI-wiki/OI-wiki.git
  HEAD branch: master
  Remote branches:
    git             tracked
    master          tracked
  ...
```

 
```

### 远程仓库的配置

执行 `git remote add <name> <url>` 命令可以添加一个名字为 `name`，链接为 `url` 的远程仓库。

执行 `git remote rename <oldname> <newname>` 可以将名字为 `oldname` 的远程仓库改名为 `newname`。

执行 `git remote rm <name>` 可以删除名字为 `name` 的远程仓库。

执行 `git remote get-url <name>` 可以查看名字为 `name` 的远程仓库的链接。

执行 `git remote set-url <name> <newurl>` 可以将名字为 `name` 的远程仓库的链接更改为 `newurl`。

### 从远程仓库获取更改

在远程仓库中，其他人可能会推送一些更改，执行 `git fetch` 命令可以将这些更改获取到本地。

```bash
$ git fetch <remote-name> # 获取 <remote-name> 的更改
```

需要注意的是，`git fetch` 命令只会获取远程仓库的更改，而不会将这些更改合并到本地仓库中。如果需要将这些更改进行合并，可以使用 `git pull` 命令。在默认情况下，`git pull` 相当于 `git fetch` 后 `git merge FETCH_HEAD`。

```bash
$ git pull <remote-name> <branch> # 获取 <remote-name> 的更改，然后将这些更改合并到 HEAD
```

### 将更改推送到远程仓库

当你完成了一些更改之后，使用 `git push` 命令可以将这些更改推送到远程仓库。

```bash
$ git push <remote> <from>:<to> # 将本地 <from> 分支的更改推送至 <remote> 的 <to> 分支
```

根据远程仓库的要求，可能会要求你输入远程仓库账户的用户名和密码。

需要注意的是，你的更改能成功推送，需要满足两个条件：你拥有向这个仓库（分支）的写入权限，且你的这个分支比远程仓库的相应分支新（可以理解为没有人在你进行更改的这段时间进行了推送）。当远程分支有当前分支没有的新更改时，可以执行 `git pull` 命令完成合并再提交。

> 如果你需要强制将本地分支的更改推送到远程仓库的话，可以加入 `-f` 参数。此时**远程仓库的提交历史会被本地的提交历史覆盖**，因此该命令应谨慎使用。

Reviewer 3: ```suggestion
如果你需要强制将本地分支的更改推送到远程仓库的话，可以加入 `-f` 参数。此时**远程仓库的提交历史会被本地的提交历史覆盖**，因此该命令应谨慎使用。更好的选择是使用 `--force-with-lease` 参数，该参数仅在远程仓库没有更新时才会进行覆盖。需要注意的是，此处“更新”是相对于上一次 fetch 而言的，如果使用了 VS Code 提供的 Auto Fetch 功能，可能会没有注意到更新而使 `--force-with-lease` 和 `-f` 一样危险。
```

### 追踪远程分支

通过将一个本地分支设定为追踪远程分支，可以方便地查看本地分支与远程分支的差别，并能简化与远程分支交互时的操作。

在开始追踪前，你需要先执行 `git fetch <remote-name>` 将远程仓库的信息抓取到本地。

接下来执行 `git switch <remote-branch>`，会在本地自动创建名字为 `<remote-branch>` 的新分支，并设定该分支自动追踪相应的远程分支。

???+note
    需要注意，只有当本地不存在该分支，且恰好只有一个远程分支的名字与该分支匹配时，Git 才会自动创建该分支且设定其追踪相应的远程分支。

这时候执行 `git status` 命令，会提示当前分支与远程分支之间的差别。

因为设定了本地分支追踪的远程分支，向远程分支推送的命令也被简化了。只需要执行 `git push` 命令，在本地分支上作出的更改就能被推送至其追踪的远程分支。

对于本地已有的分支，设定其对应的远程追踪分支也很容易。只需在当前分支下执行 `git branch -u <remote-name>/<remote-branch>`，就可以设定当前的本地分支追踪 `<remote-name>/<remote-branch>` 这一远程分支。

### 使用 ssh 连接

与 HTTP(S) 相比，使用 ssh 连接到远程仓库更为方便安全。

> 在使用 ssh 连接到远程仓库之前，需要先在本地添加 ssh 密钥。具体操作方式可以参考 [missing semester 上的教程](https://missing-semester-cn.github.io/2020/command-line/#%E8%BF%9C%E7%AB%AF%E8%AE%BE%E5%A4%87)，这里不再展开。

Reviewer 3: 我觉得这篇教程有大量连接 GitHub 所不需要了解的内容，建议还是自己写一下，可以参考 [我的 hexo 指北](https://ouuan.github.io/post/hexo%E5%8D%9A%E5%AE%A2%E6%90%AD%E5%BB%BA%E6%8C%87%E5%8C%97/#%E5%B0%86%E5%8D%9A%E5%AE%A2%E4%B8%8A%E4%BC%A0%E8%87%B3-github)，但也别直接 link 过去。

Reviewer 3: 刚刚没看清，没看到下面两行..所以你 link 这个教程就是本地添加啊..那你自己写可能比“具体操作方式可以参考，这里不再展开”字数还少一些 🌚

Reviewer 3: ```suggestion
在使用 ssh 连接到远程仓库之前，需要先在本地添加 ssh 密钥。具体来说，需要在命令行运行 `ssh-keygen` ，如果不想每次输入密码的话，不需要输入内容，连按三次回车即可。
```

Reviewer 7: > 我觉得这篇教程有大量连接 GitHub 所不需要了解的内容，建议还是自己写一下，可以参考 [我的 hexo 指北](https://ouuan.github.io/post/hexo%E5%8D%9A%E5%AE%A2%E6%90%AD%E5%BB%BA%E6%8C%87%E5%8C%97/#%E5%B0%86%E5%8D%9A%E5%AE%A2%E4%B8%8A%E4%BC%A0%E8%87%B3-github)，但也别直接 link 过去。

我认为直接链接 GitHub Doc 就好

Author: > > 我觉得这篇教程有大量连接 GitHub 所不需要了解的内容，建议还是自己写一下，可以参考 [我的 hexo 指北](https://ouuan.github.io/post/hexo%E5%8D%9A%E5%AE%A2%E6%90%AD%E5%BB%BA%E6%8C%87%E5%8C%97/#%E5%B0%86%E5%8D%9A%E5%AE%A2%E4%B8%8A%E4%BC%A0%E8%87%B3-github)，但也别直接 link 过去。
> 
> 我认为直接链接 GitHub Doc 就好

感觉还是扔个 GitHub Doc 的链接更好些。

1. 写这个页面的目的主要是给想给 OI Wiki 的各位 contributors 提供一个 guideline，让他们能用 Git 更高效的完成贡献。
2. 基于 1 的原因，他们使用的 Git 托管平台主要是 GitHub，所以扔 GitHub 的帮助会更合适些。
3. GitHub 帮助写的足够详细。

如果真的有人想对 ssh 的用法有更深入的了解的话，可以给个外部链接 :thinking:

Reviewer 3: > 感觉还是扔个 GitHub Doc 的链接更好些。
> 
>     1. 写这个页面的目的主要是给想给 OI Wiki 的各位 contributors 提供一个 guideline，让他们能用 Git 更高效的完成贡献。
> 
>     2. 基于 1 的原因，他们使用的 Git 托管平台主要是 GitHub，所以扔 GitHub 的帮助会更合适些。
> 
>     3. GitHub 帮助写的足够详细。
> 
> 
> 如果真的有人想对 ssh 的用法有更深入的了解的话，可以给个外部链接 thinking

你说的“更好”是相对什么？我觉得你给出的原因不能说明给 GitHub Doc 的链接比自己写更好。

Reviewer 3: 我是觉得对于一般人来说 GitHub Doc 的内容也过多了，按我的 suggestion 基本上已经足够了。

Reviewer 3: 而且我这直连上不了 GitHub Docs

Reviewer 7: > 我是觉得对于一般人来说 GitHub Doc 的内容也过多了，按我的 suggestion 基本上已经足够了。

GitHub Doc 给的内容就是最基本的操作步骤，我不觉得有什么多余的东西

Reviewer 7: > 而且我这直连上不了 GitHub Docs

事实上直连上不了 GitHub Doc 的 GitHub 的连接性也不怎么样，如何参与贡献呢（既然如 StudyingFather 所说本文主要目的是 contribute OI Wiki 的 guideline）

Reviewer 3: > 事实上直连上不了 GitHub Doc 的 GitHub 的连接性也不怎么样

我这就不是

Reviewer 3: > GitHub Doc 给的内容就是最基本的操作步骤，我不觉得有什么多余的东西

我觉得你的意思是“没有任何人都用不到的内容”，但我的意思是“有一个人很可能用不到的内容”。

Reviewer 3: 那你觉得

> 在使用 ssh 连接到远程仓库之前，需要先在本地添加 ssh 密钥。具体来说，需要在命令行运行 `ssh-keygen` ，如果不想每次输入密码的话，不需要输入内容，连按三次回车即可。
>
> 接下来需要将本地添加的 ssh 密钥的**公钥**上传到远程仓库。公钥可以在用户目录（Windows 上即 `C:\Users\电脑用户名`）下的 `.ssh/id_rsa.pub` 查看。
>
> 以 GitHub 为例，用户需要在 [SSH and GPG keys 设置](https://github.com/settings/keys) 中点击 **New SSH key** 来添加 ssh 密钥。

这对于（不设置密码，不更改默认保存位置的）普通用户来说缺什么吗？

> 接下来需要将本地添加的 ssh 密钥的**公钥**上传到远程仓库。以 GitHub 为例，用户需要在个人设置里点击 **SSH and GPG keys**，随后点击 **New SSH key** 来添加 ssh 密钥。

Reviewer 3: ```suggestion
接下来需要将本地添加的 ssh 密钥的**公钥**上传到远程仓库。公钥可以在用户目录（Windows 上即 `C:\Users\电脑用户名`）下的 `.ssh/id_rsa.pub` 查看。

以 GitHub 为例，用户需要在 [SSH and GPG keys 设置](https://github.com/settings/keys) 中点击 **New SSH key** 来添加 ssh 密钥。
```

Reviewer 7: 根据 `ssh-keygen` 运行时填写的内容，可能并不是这个文件名
